galsim/fitswcs.py

```diff
@@ -1761,3 +1763,92 @@ def FitsWCS(file_name=None, dir=None, hdu=None, header=None, compression='auto',
 FitsWCS._takes_rng = False
 
 
+def FittedSIPWCS(x, y, ra, dec, order=3, center=None):
+    """A WCS constructed from a list of reference celestial and image
+    coordinates.
+
+    Parameters:
+        x:      Image x-coordinates of reference stars in pixels
+        y:      Image y-coordinates of reference stars in pixels
+        ra:     Right ascension of reference stars in radians
+        dec:    Declination of reference stars in radians
+        order:  The order of the Simple Imaging Polynomial (SIP) used to
+                describe the WCS distortion.  Should be in the range [1, 9]
+                [default: 3]
+        center: A `CelestialCoord` defining the location on the sphere where the
+                tangent plane is centered.  [default: None, which means use the
+                average position of the list of reference stars]
+    """
+    from scipy.optimize import least_squares
+    from .celestial import CelestialCoord
+    if center is None:
+        # Use deprojected 3D mean of ra/dec unit sphere points as center
+        wx = np.mean(np.cos(dec)*np.cos(ra))
+        wy = np.mean(np.cos(dec)*np.sin(ra))
+        wz = np.mean(np.sin(dec))
+        center = CelestialCoord.from_xyz(wx, wy, wz)
+
+    if order < 1 or order > 9:
+        raise GalSimValueError("Illegal SIP order", order)
+
+    # Project radec onto uv so we can linearly fit the CRPIX and CD matrix
+    # initial guesses
+    u, v = center.project_rad(ra, dec)
+    a = np.array(np.broadcast_arrays(1., -u, v)).T
+    b = np.array([x, y]).T
+    r, _, _, _ = np.linalg.lstsq(a, b, rcond=None)
+    crpix_guess = r[0]
+    cd_guess = np.linalg.inv(np.deg2rad(r[1:]))
+
+    # SIP coefficient initial guesses are just 0.0
+    ab_guess = []
+    for i in range(order+1):
+        for j in range(order+1):
+            if (i+j > 1) and (i+j <= order):
+                ab_guess.extend([0.0, 0.0])
+    ab_guess = np.array(ab_guess)
+    guess = np.hstack([crpix_guess, cd_guess.ravel(), ab_guess.ravel()])
+
+    def _getWCS(params):
+        crpix = params[:2]
+        cd = params[2:6].reshape(2, 2)
+
+        k = 0
+        a = []
+        b = []
+        for i in range(order+1):
+            for j in range(order+1):
+                if (i+j < 2) or (i+j > order):
+                    a.append(0.0)
+                    b.append(0.0)
+                else:
+                    a.append(params[6+k])
+                    b.append(params[6+k+1])
+                    k += 2
+        a = np.array(a).reshape((order+1, order+1))
+        b = np.array(b).reshape((order+1, order+1))
```

Comment:
I usually like to allocate the array at the start and fill it in if possible.  But I'm not sure if this actually ends up clearer in this case.  (It seems to be irrelevant for running time, and it's not actually many fewer lines of code.)  Anyway, for your consideration, I believe the following is equivalent:
```
        a = np.zeros((order+1,order+1))
        b = np.zeros((order+1,order+1))
        k = 0
        for i in range(order+1):
            # skip i+j < 2 and i+j > order
            j1 = 0 if i > 1 else 2-i
            j2 = order + 1 - i
            a[i][j1:j2] = params[6+k : 6+k+2*(j2-j1) : 2]
            b[i][j1:j2] = params[7+k : 7+k+2*(j2-j1) : 2]
            k += 2*(j2-j1)
        ab = np.array([a, b])
```
Feel free to ignore this though and leave the code as is.

Reply:
Yeah, I like allocate-then-fill generally too.  In this case, though, I do think the current code is clearer, and as you say, not a time-sink.